X-Ray Sampling Rules :: Error reading response

### Issue
we are getting an error when X-Ray executes the Sampling Rules.

### Current Behavior
this happens every minute when our app is running. The problem occurs when processing the 'createdAt' of class SamplingRuleRecord present in GetSamplingRulesResult.
We think this may have started to happen when we started using version 2.6.0 of X-Ray SDK.

### Application Error
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 2020-06-20T16:31:09Z [Info] Successfully sent batch of 1 segments (0.015 seconds)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 20-Jun-2020 16:31:15.531 INFO [pool-5-thread-1] com.amazonaws.xray.strategy.sampling.pollers.RulePoller.pollRule Polling sampling rules.
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 20-Jun-2020 16:31:15.558 SEVERE [pool-5-thread-1] com.amazonaws.xray.strategy.sampling.pollers.RulePoller.lambda$start$0 Encountered error polling GetSamplingRules:
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 	com.amazonaws.xray.internal.XrayClientException: Error reading response.
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.amazonaws.xray.internal.UnsignedXrayClient.sendRequest(UnsignedXrayClient.java:136)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.amazonaws.xray.internal.UnsignedXrayClient.getSamplingRules(UnsignedXrayClient.java:88)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.amazonaws.xray.strategy.sampling.pollers.RulePoller.pollRule(RulePoller.java:100)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.amazonaws.xray.strategy.sampling.pollers.RulePoller.lambda$start$0(RulePoller.java:72)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at java.base/java.lang.Thread.run(Thread.java:834)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 	Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.util.Date` out of VALUE_NUMBER_FLOAT token
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5  at [Source: (sun.net.www.protocol.http.HttpURLConnection$HttpInputStream); line: 1, column: 38] (through reference chain: com.amazonaws.services.xray.model.GetSamplingRulesResult["SamplingRuleRecords"]->java.util.ArrayList[0]->com.amazonaws.services.xray.model.SamplingRuleRecord["CreatedAt"])
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1464)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1238)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1148)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.std.StdDeserializer._parseDate(StdDeserializer.java:511)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.std.DateDeserializers$DateBasedDeserializer._parseDate(DateDeserializers.java:200)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.std.DateDeserializers$DateDeserializer.deserialize(DateDeserializers.java:290)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.std.DateDeserializers$DateDeserializer.deserialize(DateDeserializers.java:273)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:156)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:285)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:156)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4482)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3471)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		at com.amazonaws.xray.internal.UnsignedXrayClient.sendRequest(UnsignedXrayClient.java:134)
/ecs/IndexerTask ecs/IndexerContainer/186e60110e7749b3b35d8ac761d56ea5 		... 9 more
/ecs/IndexerTask ecs/IndexerContainer/186

### Environment
AWS Java SDK version used: 2.13.40
AWS X-Ray SDK version used: 2.6.0
Jackson version used: 2.11.0
JDK version used: jdk11-corretto
Operating System and version: docker tomcat:9-jdk11-corretto
Webapp build and deploy to ECS Fargate Cluster with AWS Code Pipeline

--- a/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/internal/UnsignedXrayClient.java
+++ b/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/internal/UnsignedXrayClient.java
@@ -17,26 +17,35 @@ package com.amazonaws.xray.internal;
 
 import com.amazonaws.AmazonWebServiceRequest;
 import com.amazonaws.AmazonWebServiceResult;
+import com.amazonaws.SdkClientException;
 import com.amazonaws.services.xray.model.GetSamplingRulesRequest;
 import com.amazonaws.services.xray.model.GetSamplingRulesResult;
 import com.amazonaws.services.xray.model.GetSamplingTargetsRequest;
 import com.amazonaws.services.xray.model.GetSamplingTargetsResult;
 import com.amazonaws.xray.config.DaemonConfiguration;
 import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.PropertyName;
 import com.fasterxml.jackson.databind.PropertyNamingStrategy;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+import com.fasterxml.jackson.databind.introspect.Annotated;
 import com.fasterxml.jackson.databind.introspect.AnnotatedMember;
 import com.fasterxml.jackson.databind.introspect.JacksonAnnotationIntrospector;
+import com.fasterxml.jackson.databind.module.SimpleModule;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.io.UnsupportedEncodingException;
+import java.math.BigDecimal;
 import java.net.HttpURLConnection;
 import java.net.MalformedURLException;
 import java.net.ProtocolException;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
+import java.util.Date;
 import javax.annotation.Nullable;
 
 /**
@@ -47,11 +56,15 @@ import javax.annotation.Nullable;
  */
 public class UnsignedXrayClient {
 
+    private static final PropertyName HTTP_METHOD = PropertyName.construct("HTTPMethod");
+    private static final PropertyName URL_PATH = PropertyName.construct("URLPath");
+
     // Visible for testing
     static final ObjectMapper OBJECT_MAPPER = new ObjectMapper()
             .setSerializationInclusion(Include.NON_EMPTY)
             // Use deprecated field to support older Jackson versions for now.
             .setPropertyNamingStrategy(PropertyNamingStrategy.PASCAL_CASE_TO_CAMEL_CASE)
+            .registerModule(new SimpleModule().addDeserializer(Date.class, new FloatDateDeserializer()))
             .setAnnotationIntrospector(new JacksonAnnotationIntrospector() {
                 @Override
                 public boolean hasIgnoreMarker(AnnotatedMember m) {
@@ -63,6 +76,17 @@ public class UnsignedXrayClient {
                         return true;
                     }
                     return super.hasIgnoreMarker(m);
+                }
+
+                @Override
+                public PropertyName findNameForDeserialization(Annotated a) {
+                    if (a.getName().equals("hTTPMethod")) {
+                        return HTTP_METHOD;
+                    }
+                    if (a.getName().equals("uRLPath")) {
+                        return URL_PATH;
+                    }
+                    return super.findNameForDeserialization(a);
                 }
             });
     private static final int TIME_OUT_MILLIS = 2000;
@@ -165,6 +189,39 @@ public class UnsignedXrayClient {
         int b;
         while ((b = is.read()) != -1) {
             os.write(b);
+        }
+    }
+
+    private static class FloatDateDeserializer extends StdDeserializer<Date> {
+
+        private static final int AWS_DATE_MILLI_SECOND_PRECISION = 3;
+
+        private FloatDateDeserializer() {
+            super(Date.class);
+        }
+
+        @Override
+        public Date deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
+            return parseServiceSpecificDate(p.getText());
+        }
+
+        // Copied from AWS SDK https://github.com/aws/aws-sdk-java/blob/7b1e5b87b0bf03456df9e77716b14731adf9a7a7/aws-java-sdk-core/src/main/java/com/amazonaws/util/DateUtils.java#L239https://github.com/aws/aws-sdk-java/blob/7b1e5b87b0bf03456df9e77716b14731adf9a7a7/aws-java-sdk-core/src/main/java/com/amazonaws/util/DateUtils.java#L239
+        /**
+         * Parses the given date string returned by the AWS service into a Date
+         * object.
+         */
+        private static Date parseServiceSpecificDate(String dateString) {
+            if (dateString == null) {
+                return null;
+            }
+            try {
+                BigDecimal dateValue = new BigDecimal(dateString);
+                return new Date(dateValue.scaleByPowerOfTen(
+                    AWS_DATE_MILLI_SECOND_PRECISION).longValue());
+            } catch (NumberFormatException nfe) {
+                throw new SdkClientException("Unable to parse date : "
+                                             + dateString, nfe);
+            }
         }
     }
 }

--- a/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/internal/UnsignedXrayClient.java
+++ b/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/internal/UnsignedXrayClient.java
@@ -102,7 +102,7 @@ public class UnsignedXrayClient {
     UnsignedXrayClient(String endpoint) {
         try {
             getSamplingRulesEndpoint = new URL(endpoint + "/GetSamplingRules");
-            getSamplingTargetsEndpoint = new URL(endpoint + "/GetSamplingTargets");
+            getSamplingTargetsEndpoint = new URL(endpoint + "/SamplingTargets");
         } catch (MalformedURLException e) {
             throw new IllegalArgumentException("Invalid URL: " + endpoint, e);
         }

--- a/aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/internal/UnsignedXrayClientTest.java
+++ b/aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/internal/UnsignedXrayClientTest.java
@@ -245,7 +245,7 @@ public class UnsignedXrayClientTest {
         GetSamplingTargetsResult expected = OBJECT_MAPPER.readValue(SAMPLING_TARGETS, GetSamplingTargetsResult.class);
         assertThat(expected).isEqualTo(result);
 
-        verify(postRequestedFor(urlEqualTo("/GetSamplingTargets"))
+        verify(postRequestedFor(urlEqualTo("/SamplingTargets"))
                        .withHeader("Content-Type", equalTo("application/json"))
                        .withRequestBody(equalToJson("{"
                                                     + " \"SamplingStatisticsDocuments\": ["

--- a/aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/internal/UnsignedXrayClientTest.java
+++ b/aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/internal/UnsignedXrayClientTest.java
@@ -35,7 +35,6 @@ import com.amazonaws.services.xray.model.GetSamplingTargetsRequest;
 import com.amazonaws.services.xray.model.GetSamplingTargetsResult;
 import com.amazonaws.services.xray.model.SamplingStatisticsDocument;
 import com.github.tomakehurst.wiremock.junit.WireMockClassRule;
-import java.util.Date;
 import org.junit.Before;
 import org.junit.ClassRule;
 import org.junit.Test;
@@ -44,6 +43,170 @@ public class UnsignedXrayClientTest {
 
     @ClassRule
     public static WireMockClassRule server = new WireMockClassRule(wireMockConfig().dynamicPort());
+
+    private static final String SAMPLING_RULES =
+        "{\n"
+        + "    \"SamplingRuleRecords\": [\n"
+        + "        {\n"
+        + "            \"SamplingRule\": {\n"
+        + "                \"RuleName\": \"Default\",\n"
+        + "                \"RuleARN\": \"arn:aws:xray:us-east-1::sampling-rule/Default\",\n"
+        + "                \"ResourceARN\": \"*\",\n"
+        + "                \"Priority\": 10000,\n"
+        + "                \"FixedRate\": 0.01,\n"
+        + "                \"ReservoirSize\": 0,\n"
+        + "                \"ServiceName\": \"*\",\n"
+        + "                \"ServiceType\": \"*\",\n"
+        + "                \"Host\": \"*\",\n"
+        + "                \"HTTPMethod\": \"*\",\n"
+        + "                \"URLPath\": \"*\",\n"
+        + "                \"Version\": 1,\n"
+        + "                \"Attributes\": {}\n"
+        + "            },\n"
+        + "            \"CreatedAt\": 0.0,\n"
+        + "            \"ModifiedAt\": 1530558121.0\n"
+        + "        },\n"
+        + "        {\n"
+        + "            \"SamplingRule\": {\n"
+        + "                \"RuleName\": \"base-scorekeep\",\n"
+        + "                \"RuleARN\": \"arn:aws:xray:us-east-1::sampling-rule/base-scorekeep\",\n"
+        + "                \"ResourceARN\": \"*\",\n"
+        + "                \"Priority\": 9000,\n"
+        + "                \"FixedRate\": 0.1,\n"
+        + "                \"ReservoirSize\": 2,\n"
+        + "                \"ServiceName\": \"Scorekeep\",\n"
+        + "                \"ServiceType\": \"*\",\n"
+        + "                \"Host\": \"*\",\n"
+        + "                \"HTTPMethod\": \"*\",\n"
+        + "                \"URLPath\": \"*\",\n"
+        + "                \"Version\": 1,\n"
+        + "                \"Attributes\": {}\n"
+        + "            },\n"
+        + "            \"CreatedAt\": 1530573954.0,\n"
+        + "            \"ModifiedAt\": 1530920505.0\n"
+        + "        },\n"
+        + "        {\n"
+        + "            \"SamplingRule\": {\n"
+        + "                \"RuleName\": \"polling-scorekeep\",\n"
+        + "                \"RuleARN\": \"arn:aws:xray:us-east-1::sampling-rule/polling-scorekeep\",\n"
+        + "                \"ResourceARN\": \"*\",\n"
+        + "                \"Priority\": 5000,\n"
+        + "                \"FixedRate\": 0.003,\n"
+        + "                \"ReservoirSize\": 0,\n"
+        + "                \"ServiceName\": \"Scorekeep\",\n"
+        + "                \"ServiceType\": \"*\",\n"
+        + "                \"Host\": \"*\",\n"
+        + "                \"HTTPMethod\": \"GET\",\n"
+        + "                \"URLPath\": \"/api/state/*\",\n"
+        + "                \"Version\": 1,\n"
+        + "                \"Attributes\": {}\n"
+        + "            },\n"
+        + "            \"CreatedAt\": 1530918163.0,\n"
+        + "            \"ModifiedAt\": 1530918163.0\n"
+        + "        }\n"
+        + "    ]\n"
+        + "}{\n"
+        + "    \"SamplingRuleRecords\": [\n"
+        + "        {\n"
+        + "            \"SamplingRule\": {\n"
+        + "                \"RuleName\": \"Default\",\n"
+        + "                \"RuleARN\": \"arn:aws:xray:us-east-1::sampling-rule/Default\",\n"
+        + "                \"ResourceARN\": \"*\",\n"
+        + "                \"Priority\": 10000,\n"
+        + "                \"FixedRate\": 0.01,\n"
+        + "                \"ReservoirSize\": 0,\n"
+        + "                \"ServiceName\": \"*\",\n"
+        + "                \"ServiceType\": \"*\",\n"
+        + "                \"Host\": \"*\",\n"
+        + "                \"HTTPMethod\": \"*\",\n"
+        + "                \"URLPath\": \"*\",\n"
+        + "                \"Version\": 1,\n"
+        + "                \"Attributes\": {}\n"
+        + "            },\n"
+        + "            \"CreatedAt\": 0.0,\n"
+        + "            \"ModifiedAt\": 1530558121.0\n"
+        + "        },\n"
+        + "        {\n"
+        + "            \"SamplingRule\": {\n"
+        + "                \"RuleName\": \"base-scorekeep\",\n"
+        + "                \"RuleARN\": \"arn:aws:xray:us-east-1::sampling-rule/base-scorekeep\",\n"
+        + "                \"ResourceARN\": \"*\",\n"
+        + "                \"Priority\": 9000,\n"
+        + "                \"FixedRate\": 0.1,\n"
+        + "                \"ReservoirSize\": 2,\n"
+        + "                \"ServiceName\": \"Scorekeep\",\n"
+        + "                \"ServiceType\": \"*\",\n"
+        + "                \"Host\": \"*\",\n"
+        + "                \"HTTPMethod\": \"*\",\n"
+        + "                \"URLPath\": \"*\",\n"
+        + "                \"Version\": 1,\n"
+        + "                \"Attributes\": {}\n"
+        + "            },\n"
+        + "            \"CreatedAt\": 1530573954.0,\n"
+        + "            \"ModifiedAt\": 1530920505.0\n"
+        + "        },\n"
+        + "        {\n"
+        + "            \"SamplingRule\": {\n"
+        + "                \"RuleName\": \"polling-scorekeep\",\n"
+        + "                \"RuleARN\": \"arn:aws:xray:us-east-1::sampling-rule/polling-scorekeep\",\n"
+        + "                \"ResourceARN\": \"*\",\n"
+        + "                \"Priority\": 5000,\n"
+        + "                \"FixedRate\": 0.003,\n"
+        + "                \"ReservoirSize\": 0,\n"
+        + "                \"ServiceName\": \"Scorekeep\",\n"
+        + "                \"ServiceType\": \"*\",\n"
+        + "                \"Host\": \"*\",\n"
+        + "                \"HTTPMethod\": \"GET\",\n"
+        + "                \"URLPath\": \"/api/state/*\",\n"
+        + "                \"Version\": 1,\n"
+        + "                \"Attributes\": {}\n"
+        + "            },\n"
+        + "            \"CreatedAt\": 1530918163.0,\n"
+        + "            \"ModifiedAt\": 1530918163.0\n"
+        + "        }\n"
+        + "    ]\n"
+        + "}";
+
+    private static final String SAMPLING_TARGETS =
+        "{\n"
+        + "    \"SamplingTargetDocuments\": [\n"
+        + "        {\n"
+        + "            \"RuleName\": \"base-scorekeep\",\n"
+        + "            \"FixedRate\": 0.1,\n"
+        + "            \"ReservoirQuota\": 2,\n"
+        + "            \"ReservoirQuotaTTL\": 1530923107.0,\n"
+        + "            \"Interval\": 10\n"
+        + "        },\n"
+        + "        {\n"
+        + "            \"RuleName\": \"polling-scorekeep\",\n"
+        + "            \"FixedRate\": 0.003,\n"
+        + "            \"ReservoirQuota\": 0,\n"
+        + "            \"ReservoirQuotaTTL\": 1530923107.0,\n"
+        + "            \"Interval\": 10\n"
+        + "        }\n"
+        + "    ],\n"
+        + "    \"LastRuleModification\": 1530920505.0,\n"
+        + "    \"UnprocessedStatistics\": []\n"
+        + "}{\n"
+        + "    \"SamplingTargetDocuments\": [\n"
+        + "        {\n"
+        + "            \"RuleName\": \"base-scorekeep\",\n"
+        + "            \"FixedRate\": 0.1,\n"
+        + "            \"ReservoirQuota\": 2,\n"
+        + "            \"ReservoirQuotaTTL\": 1530923107.0,\n"
+        + "            \"Interval\": 10\n"
+        + "        },\n"
+        + "        {\n"
+        + "            \"RuleName\": \"polling-scorekeep\",\n"
+        + "            \"FixedRate\": 0.003,\n"
+        + "            \"ReservoirQuota\": 0,\n"
+        + "            \"ReservoirQuotaTTL\": 1530923107.0,\n"
+        + "            \"Interval\": 10\n"
+        + "        }\n"
+        + "    ],\n"
+        + "    \"LastRuleModification\": 1530920505.0,\n"
+        + "    \"UnprocessedStatistics\": []\n"
+        + "}";
 
     private UnsignedXrayClient client;
 
@@ -54,15 +217,13 @@ public class UnsignedXrayClientTest {
 
     @Test
     public void getSamplingRules() throws Exception {
-        GetSamplingRulesResult expected = new GetSamplingRulesResult();
-        expected.setNextToken("nexttoken");
-
         stubFor(any(anyUrl()).willReturn(aResponse()
                                                  .withStatus(200)
-                                                 .withBody(OBJECT_MAPPER.writeValueAsBytes(expected))));
+                                                 .withBody(SAMPLING_RULES)));
 
         GetSamplingRulesResult result = client.getSamplingRules(new GetSamplingRulesRequest());
 
+        GetSamplingRulesResult expected = OBJECT_MAPPER.readValue(SAMPLING_RULES, GetSamplingRulesResult.class);
         assertThat(expected).isEqualTo(result);
 
         verify(postRequestedFor(urlEqualTo("/GetSamplingRules"))
@@ -72,18 +233,16 @@ public class UnsignedXrayClientTest {
 
     @Test
     public void getSamplingTargets() throws Exception {
-        GetSamplingTargetsResult expected = new GetSamplingTargetsResult();
-        expected.setLastRuleModification(new Date(1000));
-
         stubFor(any(anyUrl()).willReturn(aResponse()
                                                  .withStatus(200)
-                                                 .withBody(OBJECT_MAPPER.writeValueAsBytes(expected))));
+                                                 .withBody(SAMPLING_TARGETS)));
 
         GetSamplingTargetsRequest request = new GetSamplingTargetsRequest()
             .withSamplingStatisticsDocuments(new SamplingStatisticsDocument().withClientID("client-id"));
 
         GetSamplingTargetsResult result = client.getSamplingTargets(request);
 
+        GetSamplingTargetsResult expected = OBJECT_MAPPER.readValue(SAMPLING_TARGETS, GetSamplingTargetsResult.class);
         assertThat(expected).isEqualTo(result);
 
         verify(postRequestedFor(urlEqualTo("/GetSamplingTargets"))

